src/csharp/Grpc.Tools/DepFileUtil.cs

```diff
@@ -269,5 +301,51 @@ static int FindLineSeparator(string line)
                 return new string[0];
             }
         }
+
+        // Calculate part of proto path relative to root. Protoc is very picky
+        // about them matching exactly, so can be we. Expect root be exact prefix
+        // to proto, minus some slash normalization.
+        internal static string GetRelativeDir(string root, string proto, TaskLoggingHelper log)
+        {
+            string protoDir = Path.GetDirectoryName(proto);
+            string rootDir = EndWithSlash(Path.GetDirectoryName(EndWithSlash(root)));
+            if (rootDir == s_dotSlash)
+            {
+                // Special case, otherwise we can return "./" instead of "" below!
+                return protoDir;
+            }
+            if (Platform.IsFsCaseInsensitive)
+            {
+                protoDir = protoDir.ToLowerInvariant();
+                rootDir = rootDir.ToLowerInvariant();
+            }
+            protoDir = EndWithSlash(protoDir);
+            if (!protoDir.StartsWith(rootDir))
+            {
+                log.LogWarning("Protobuf item '{0}' has the ProtoRoot metadata '{1}' " +
+                               "which is not prefix to its path. Cannot compute relative path.",
+                    proto, root);
+                return "";
+            }
+            return protoDir.Substring(rootDir.Length);
+        }
+
+        // './' or '.\', normalized per system.
+        internal static string s_dotSlash = "." + Path.DirectorySeparatorChar;
+
+        internal static string EndWithSlash(string str)
```

Comment:
nit: `EnsureTrailingSlash` sound more descriptive for the purpose of this method.

Reply:
Only moved it from GeneratorServices. Will keep it there and rename it.